docs/rfc/0000-json-schema-validation.md

```diff
@@ -0,0 +1,272 @@
+```
+
+## Motivation
+
+`jk` and its libraries provide various means for generating
+configuration, but there is in general no guarantee that the output is
+usable. For that reason, to become a more rounded tool, validation
+should be considered a core capability for `jk`.
+
+One very general-purpose means of validation is by checking against a
+data model or schema. In the case of JSON-serialisable JavaScript
+objects -- which is effectively the data model of `jk` -- a reasonable
+choice of schema language is JSON Schema. Schemas using JSON Schema
+(and the closely related OpenAPI data model, more on which see below)
+are available for lots of APIs, and it has support in tooling e.g.,
+[VSCode](https://code.visualstudio.com/docs/languages/json#_json-schemas-and-settings).
+
+Building in schema validation means libraries are free to rely on it,
+encouraging more widespread validation.
+
+## Design
+
+### Discussion
+
+**Synchronisation**
+
+The expectation might be that validation is synchronous, since it is
+used mostwhere as a predicate. There's good reasons why it should be
+asynchronous, though:
+
+ - validating with a file means a file read, and those are
+   asynchronous, so the precedent exists;
+ - resolving references may involve network requests, and those are
+   naturally represented as asynchronous.
+
+**Resolving schemas**
+
+ - `$schema` values will likely refer to public URIs (but this RFC
+   says nothing about respecting `$schema`; that could follow in
+   another RFC, perhaps).
+ - schemas will often refer to other schemas with `$ref: <json
+   pointer>`, which can come from the same file, or from a file
+   relative to the current file, or a URI (though it's not required to
+   be a network locator, i.e., it's up to the library how to process
+   it)
+ - libraries will want to include their schemata in the distributed
+   package, so will want to be able to load them as a resource.
+
+**Caching**
+
+Usually, caching is strictly an optimisation, and as such it is not
+necessary to discuss it by way of specification. But it's worth
+mentioning that Go libraries for schema checking tend to come with a
+cache for external schemas; and some schemas will be used extensively
+during a run (e.g., the schema for a Kubernetes Deployment). It would
+be good to work in sympathy with any caching in the library, where
+possible.
+
+### API in JavaScript
+
+The result of a schema check is either that everything was OK, or that
+there were some specific problems.
+
+```typescript
+type ValidationResult = 'ok' | string[];
+```
+
+Either `'ok'` or `string[]` indicates a successful validation call, so
+either can appear as the resolution to a promise. A promise will be
+rejected if there's an error in the validation process itself -- for
+example, a file cannot be found.
+
+The obvious mode of use is to supply the value and the schema, as
+JavaScript objects:
+
+```typescript
+validateBySchema: (value: any, schemaObj: any) => ValidationResult;
```

Comment:
Wondering if `validateWithSchema` would be better than `By`

Reply:
I oscillated between `validateWithObject`, `validateBySchemaObject`, `validateWithSchema` as I wrote the RFC then the code :-) Of those:, including `schema` is probably redundant since that's the module name; `validate` is probably _not_ redundant, since we may end up putting other things in there (e..g, constructors for object schema); `With` and `By` are similar, but I can certainly be persuaded that `With` edges it.

```
import * as schema from '@jkcfg/std/schema';

const defn = { type: 'number' };

function validate(obj) {
  return schema.validateWithObject(obj, defn);
}
```

This looks fine ^

Reply:
This looks good to me too.

Reply:
Made it `{ validateWithObject, validateWithFile, validateWithResource }`